man/xfardic.1
.TH xFarDic-Multilingual-Dictionary 1 "April 26 2008" "Release: 0.11.1"
.SH NAME
    xfardic - xFarDic Multilingual Dictionary 
.SH DESCRIPTION

.B What's xFarDic

.BxFarDic
is a multilingual dictionary software developed using\fB wxWidgets\fR, \fBGTK2\fR and \fBlibxml2\fR for GNU/Linux, GNOME/XFce4/KDE environments. Generic and electeronic English-Persian (Farsi) and GNU/Linux English-English word databases are available to download. xFarDic's word database format is plain XML format. xFarDic is free software released under the GNU/GPL. 

.SH HOME PAGE
Visit \fBxFarDic\fR  projects's homepage at: \fIhttp://www.xfardic.org\fR. For more technichal information you can visit xFarDic's Sourceforge.net page at: \fIhttp://sourceforge.net/projects/xfardic\fR.

.SH FEATURES
See xFarDic features on the \fIhttp://www.xfardic.org/html/module-htmlpages-display-pid-3.html\fR page.

.SH USER'S GUIDE
xFarDic User's Guide can help you to install use and tweak your xFarDic much more easier. Read it on the \fIhttp://www.xfardic.org/html/module-htmlpages-display-pid-1.html\fR page.

.SH LATEST NEWS
You can get latest news about \fBxFarDic\fR from project's homepage at: \fIhttp://www.xfardic.org\fR. If you need rss feed of Parsix GNU/Linux's news page use: \fIhttp://www.xfardic.org/html/backend.php\fR.

.Sh IRC Channel
Join to xFarDic's IRC channel at \fIirc.freenode.net/#technotux\fR. You can use any IRC client program such as xChat to use IRC service.

.SH COPYRIGHT
\fBxFarDic\fR is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version. xFarDic is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.  

.SH NOTE
\fBxFarDic\fR is packaged without any databases. You can get required databases from \fIhttp://sourceforge.net/project/showfiles.php?group_id=103427&package_id=165900\fR. Without database, xFarDic *will not* work.

.SH FILES
 
.TP
.I  /usr/share/xfardic/locales or /usr/local/share/xfardic/locales
Locale files for xFarDic Multilingual Dictionary
.P
.TP
.I /usr/share/xfardic/pixmaps or /usr/local/share/xfardic/pixmaps
Pixmaps for xFarDic Multilingual Dictionary
.P
.TP
.I /usr/bin/xfardic or /usr/local/bin/xfardic
xFarDic binary fie
.P 
.TP
.I  ~/.xfardic 
xFarDic user configuration and settings, which is an ASCII file.

.SH REPORTING BUGS
Help us to make xFarDic better. To report bugs and feature requests, send email to: \fIalan@technotux.org\fR      
		 		
.SH AUTHORS
Alan Baghumian <a\fIlan@technotux.org\fR> and Armen Baghumian <\fIarmen@opensourceclub.org\fR>
.PP
This Manual page is written by Navid Abdi <\fInavid@gnuiran.org\fR>. 


NEWS
--------------------------------------------------------------
2006-12-08						  NEWS
--------------------------------------------------------------
xFarDic 0.8.1 released. This is a patch release with two new
features. The ProcessWord() function and new about window. Also
there are few fixes on Make files.
--------------------------------------------------------------
2006-12-20						  NEWS
--------------------------------------------------------------
xFarDic 0.8.2 released. A bug fix and minor-new feature release.
Fixed a bug on the ProcessWord() function. New settings window.
--------------------------------------------------------------
2006-12-28						  NEWS
--------------------------------------------------------------
xFarDic 0.8.3 released. This is a bug fix release to solve lost
focus bug on input box and translate button. Also there are some
updates on about and settings window and an improved logo. Also
applied 16x16 icons on menus and pop-up window.
--------------------------------------------------------------
2007-01-26						  NEWS
--------------------------------------------------------------
xFarDic 0.8.4 released. This is a small feature update and bug
fix release that adds 'ment' and 'er' support to the ProcessWord 
function and also adds separate option to enable/disable word 
selection scanning system. Also there was some fixes on the 
Make files to install svg logo at the proper location and to
check xdbs existence before parsing and to remove missing 
xdbs automatically.
--------------------------------------------------------------
2007-02-XX						  NEWS
--------------------------------------------------------------
xFarDic 0.9.0 released. This is a major upgrade to the xFarDic
features and introduces leitner box a tool to help memorizing
words in an efficient way. Other changes include new icon for
clear history function and parenthesis support for the Process
word function.


src/xfardic.cpp
// +-------------------------------------------------------------------------+
// | xFarDic Multilingual Dictionary                                         |
// | Copyright (C) 2004-2008 by the xFarDic Development Team.                |
// | http://www.xfardic.org/                                                 |
// +-------------------------------------------------------------------------+
// | License:                                                                |
// |                                                                         |
// | xFarDic is free software; you can redistribute it and/or modify         |
// | it under the terms of the GNU General Public License as published by    |
// | the Free Software Foundation; either version 2 of the License, or       |
// | (at your option) any later version.                                     |
// |                                                                         |
// | xFarDic is distributed in the hope that it will be useful,              |
// | but WITHOUT ANY WARRANTY; without even the implied warranty of          |
// | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the            |
// | GNU General Public License for more details.                            |
// |                                                                         |
// | You should have received a copy of the GNU General Public License       |
// | along with xFarDic; if not, write to the Free Software                  |
// | Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA|
// |                                                                         |
// +-------------------------------------------------------------------------+
// | Authors: Alan Baghumian <alan@technotux.org                             |
// |          Armen Baghumian <armen@OpenSourceClub.org>                     |
// +-------------------------------------------------------------------------+

#include "my_app.h"

// Create a new application object: this macro will allow wxWindows to create
// the application object during program execution (it's better than using a
// static object for many reasons) and also declares the accessor function
// wxGetApp() which will return the reference of the right type (i.e. MyApp and
// not wxApp)

IMPLEMENT_APP(MyApp)

BEGIN_EVENT_TABLE(MyApp, wxApp)
    EVT_TASKBAR_RIGHT_UP(MyApp::OnTaskRightDown)
    EVT_TASKBAR_LEFT_UP(MyApp::OnTaskLeftDown)
    EVT_MENU(xFarDic_QApp,  MyApp::OnQuit)    
    EVT_MENU(xFarDic_Paste,  MyApp::OnPaste)    
    EVT_MENU(xFarDic_About,  MyApp::OnAbout)    
    EVT_MENU(xFarDic_Settings,  MyApp::OnSettings)    
    EVT_MENU(xFarDic_Setfont,  MyApp::OnSetFont)    
END_EVENT_TABLE()

/// 'Main program' equivalent: the program execution "starts" here
bool MyApp::OnInit()
{
    // Disabling wxLog
    wxLogNull logNo;

    wxString logo, Param;

    // Command line argument
    static const wxCmdLineEntryDesc cmdLineDesc[] =
    {
        { wxCMD_LINE_SWITCH, _T("h"), _T("help"), _("displays this help message"), wxCMD_LINE_VAL_NONE, wxCMD_LINE_OPTION_HELP },
        { wxCMD_LINE_PARAM, NULL, NULL, wxT("word"), wxCMD_LINE_VAL_STRING, wxCMD_LINE_PARAM_OPTIONAL },
        { wxCMD_LINE_NONE }
    };
    wxCmdLineParser parser(cmdLineDesc, wxApp::argc, wxApp::argv);
    logo.Printf(wxT("xFarDic %s (C)2004-2007 Alan Baghumian, Armen Baghumian\nxFarDic is free software released under the GPL.\n"),
          XVERSION);
    parser.SetLogo(logo);
    
    if (parser.Parse()) {
        return FALSE;
    }

    Param = parser.GetParam(0);
    
    // DEBUGGING
    // fprintf(stderr, "%s\n", (const char *)Param.mb_str(wxConvUTF8));
    // End Command line argument

    wxString xapp = wxString::Format(wxT("xFarDic"), wxGetUserId().c_str());
    m_checker = new wxSingleInstanceChecker(xapp);
    if (m_checker->IsAnotherRunning()) {
        wxString msg;
        msg.Printf( _("Another program instance is already running, aborting."));
        wxMessageBox(msg, _T("xFarDic"), wxOK | wxICON_EXCLAMATION, 0);
        return FALSE;
    }

    wxImage::AddHandler(new wxPNGHandler);

    wxConfigBase *pConfig = wxConfigBase::Get();

    // note that it makes no sense to translate these strings, they are
    // shown before we set the locale anyhow
    const wxString langNames[] =
    {
        _T("Farsi"),
        _T("English (U.S.)"),
        _T("System default"),
        //_T("Azeri")
    };

    static const wxLanguage langIds[] =
    {
        wxLANGUAGE_FARSI,
        wxLANGUAGE_ENGLISH,
        wxLANGUAGE_DEFAULT,
        //wxLANGUAGE_AZERI_LATIN
    };

    // the arrays should be in sync
    wxCOMPILE_TIME_ASSERT( WXSIZEOF(langNames) == WXSIZEOF(langIds),
                           LangArraysMismatch );

    pConfig->SetPath(wxT("/Options"));
    m_locale.Init(langIds[pConfig->Read(_T("GUI-Lang"), 2)]);  
    
    // m_locale.Init(langIds[1]);

    // Initialize the catalogs we'll be using
    m_locale.AddCatalogLookupPathPrefix(wxT("/usr/share/locale/"));
    m_locale.AddCatalogLookupPathPrefix(wxT("/usr/local/share/locale/"));
    m_locale.AddCatalog(_T("xfardic"));

    // this catalog is installed in standard location on Linux systems and
    // shows that you may make use of the standard message catalogs as well
    //
    // if it's not installed on your system, it is just silently ignored
    wxLogNull noLog;
    m_locale.AddCatalog(_T("fileutils"));

    //DEBUGGING
    //fprintf(stderr, "Going to starting Frame\n");   

    // restore frame position & state
    int x = pConfig->Read(_T("x"), 0l);
    int y = pConfig->Read(_T("y"), 0l);
    int w = pConfig->Read(_T("w"), 0l);
    int h = pConfig->Read(_T("h"), 0l);
    bool maximized = pConfig->Read(_T("IsMaximized"), 0l);

    // create the main application window
    if (h > 0 && w > 0) {
         frame = new xFarDicApp(_T("xFarDic ")XVERSION, wxPoint(x,y), wxSize(w,h), m_locale,
                       wxDEFAULT_FRAME_STYLE, FALSE);
    } else {
         frame = new xFarDicApp(_T("xFarDic ")XVERSION, wxPoint(x,y), wxDefaultSize, m_locale,
                       wxDEFAULT_FRAME_STYLE, TRUE);
    }

    // Set minimum allowed size
    frame->SetMinSize(wxSize(509, 294));

    //DEBUGGING
    //fprintf(stderr, "Frame Started\n");    
                 
    if (x == 0 && y == 0) {
        frame->Centre();
    }

    if (maximized) {
        frame->Maximize();
    }

    // translate sent argument or first cached word
    int saved = pConfig->Read(_T("Save-Cache"), 1);

    if (Param.Len() != 0 && frame->CheckSpell(Param,0)) {
        pConfig->Write(wxT("/Options/Temp-String"), Param);
    } else {
        if (saved != 0) {
            wxString str;
            long dummy;
            pConfig->SetPath(wxT("/Cache"));
            bool bCont = pConfig->GetFirstEntry(str, dummy);
            if (bCont) {
                frame->translate();
            }
        }
    } 

    // and show it (the frames, unlike simple controls, are not shown when
    // created initially)
      
    frame->Show(TRUE);
    
    // success: wxApp::OnRun() will be called which will enter the main message
    // loop and the application will run. If we returned FALSE here, the
    // application would exit immediately.
    
    // DEBUGGING
    //fprintf(stderr, "Finished OnInit()\n");
    return TRUE;
}

/// Taskbar right-click event handler
void MyApp::OnTaskRightDown(wxTaskBarIconEvent& WXUNUSED(event))
{           
     frame->PopMenu();    
     //DEBUGGING
     //fprintf(stderr, "task right click\n");
}

/// Taskbar left-click event handler
void MyApp::OnTaskLeftDown(wxTaskBarIconEvent& WXUNUSED(event))
{

    if (frame->IsIconized()) {
        if (frame->hide) {
            frame->Show(TRUE);
            if (!showLeitner) {
               frame->ltframe->Show(TRUE);
            }
            if (!showAbout) {
               frame->abframe->Show(TRUE);
            }
            if (!showSettings) {
               frame->setframe->Show(TRUE);
            }
        }
        frame->Raise();
        frame->SetFocus();                                                                                    
    } else {
        if (frame->hide) {
            frame->Show(FALSE);
            if (!showLeitner) {
               frame->ltframe->Show(FALSE);
            }
            if (!showAbout) {
               frame->abframe->Show(FALSE);
            }
            if (!showSettings) {
               frame->setframe->Show(FALSE);
            }
        }else {
            frame->Iconize();
        }
    }                                     
    //DEBUGGING
    //fprintf(stderr, "task left click\n");
}

void MyApp::OnSetFont(wxCommandEvent& WXUNUSED(event))
{
    frame->Show(TRUE);
    frame->Raise();
    frame->SetFocus();
    frame->ShowSetFont();
}

void MyApp::OnSettings(wxCommandEvent& WXUNUSED(event))
{
    frame->Show(TRUE);
    frame->Raise();
    frame->SetFocus();
    frame->ShowSettings();
}

void MyApp::OnAbout(wxCommandEvent& WXUNUSED(event))
{
    frame->Show(TRUE);        
    frame->Raise();
    frame->SetFocus();
    frame->ShowAbout();
}

void MyApp::OnPaste(wxCommandEvent& WXUNUSED(event))
{
  
    frame->Show(TRUE);
    frame->Raise();
    frame->SetFocus();
    frame->DoPaste();
}

void MyApp::OnQuit(wxCommandEvent& WXUNUSED(event))
{           
    frame->Destroy();  
    //DEBUGGING
    //fprintf(stderr, "Quit\n");
}

MyApp::~MyApp()
{
    // Delete execution Lock file
    delete m_checker;
}

